File: src/main/java/me/andre111/d20server/scripting/expression/Dice.java
package me.andre111.d20server.scripting.expression;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import me.andre111.d20common.model.entity.map.Map;
import me.andre111.d20common.model.entity.profile.Profile;
import me.andre111.d20server.scripting.ScriptException;
import me.andre111.d20server.util.DiceRoller;

public class Dice implements Expression {
	private final int count;
	private final int sides;

	private Condition criticalFailureCondition;
	private Condition criticalSuccessCondition;
	
	private Condition rerollCondition;
	private int maxRerollCount = 100;
	
	private boolean explodeDice = false;
	private int maxExplodeCount = 100;
	
	private Action lowAction = Action.NOTHING;
	private int lowCount = 0;
	private Action highAction = Action.NOTHING;
	private int highCount = 0;
	
	public Dice(int count, int sides) {
		this.count = count;
		this.sides = sides;

		this.criticalFailureCondition = new Condition(Condition.Type.EQUAL, 1);
		this.criticalSuccessCondition = new Condition(Condition.Type.EQUAL, sides);
	}
	
	public void setCriticalFailureCondition(Condition criticalFailureCondition) {
		this.criticalFailureCondition = criticalFailureCondition;
	}

	public void setCriticalSuccessCondition(Condition criticalSuccessCondition) {
		this.criticalSuccessCondition = criticalSuccessCondition;
	}

	public void setRerollCondition(Condition rerollCondition) {
		this.rerollCondition = rerollCondition;
	}

	public void setMaxRerollCount(int maxRerollCount) {
		this.maxRerollCount = maxRerollCount;
	}

	public void setExplodeDice(boolean explodeDice) {
		this.explodeDice = explodeDice;
	}

	public void setMaxExplodeCount(int maxExplodeCount) {
		this.maxExplodeCount = maxExplodeCount;
	}

	public void setLowAction(Action lowAction) {
		this.lowAction = lowAction;
	}

	public void setLowCount(int lowCount) {
		this.lowCount = lowCount;
	}

	public void setHighAction(Action highAction) {
		this.highAction = highAction;
	}

	public void setHighCount(int highCount) {
		this.highCount = highCount;
	}

	@Override
	public Result eval(Map map, Profile profile) throws ScriptException {
		// calculate all required dice
		List<DiceResult> results = new ArrayList<>();
		for(int i=0; i<count; i++) {
			int roll;
			int explodeCount = 0;
			do {
				// calculate roll
				roll = DiceRoller.roll(sides);
	
				// handle re-rolling
				int rerollCount = 0;
				while(rerollCondition != null && rerollCondition.matches(roll) && rerollCount < maxRerollCount) {
					results.add(new DiceResult(roll, false));
					roll = DiceRoller.roll(sides);
					rerollCount++;
				}
				
				results.add(new DiceResult(roll, true));
			// handle exploding dice
			} while(explodeDice && roll == sides && explodeCount++ < maxExplodeCount);
		}
		
		// apply drop/keep modifiers
		List<DiceResult> sortedResults = new ArrayList<>(results);
		Collections.sort(sortedResults);
		int currentLowCount = 0;
		for(int i=0; i<sortedResults.size(); i++) {
			DiceResult result = sortedResults.get(i);
			if(result.counted) {
				if(lowAction == Action.DROP && currentLowCount < lowCount) {
					result.dropped = true;
				} else if(lowAction == Action.KEEP && currentLowCount >= lowCount) {
					result.kept = false;
				}
				currentLowCount++;
			}
		}
		int currentHighCount = 0;
		for(int i=sortedResults.size()-1; i>=0; i--) {
			DiceResult result = sortedResults.get(i);
			if(result.counted) {
				if(highAction == Action.DROP && currentHighCount < highCount) {
					result.dropped = true;
				} else if(highAction == Action.KEEP && currentHighCount >= highCount) {
					result.kept = false;
				}
				currentHighCount++;
			}
		}
		
		// calculate result
		int value = 0;
		StringBuilder sb = new StringBuilder();
		boolean hadCriticalSuccess = false;
		boolean hadCriticalFailure = false;
		if(results.size() > 1) sb.append("(");
		
		for(int i=0; i<results.size(); i++) {
			DiceResult result = results.get(i);
			// apply value
			if(result.shouldBeIncluded()) {
				value += result.value;
				if(criticalFailureCondition.matches(value)) hadCriticalFailure = true;
				else if(criticalSuccessCondition.matches(value)) hadCriticalSuccess = true;
			}
			
			// build string (with colored misses/crits)
			if(i > 0) {
				sb.append("+");
			}
			sb.append("[group \"order=STACKED;align-vertical=CENTER;align-horizontal=CENTER\"");
			//TODO: better display of uncounted dice
			if(!result.shouldBeIncluded()) sb.append("-----");
			if(sides == 4 || sides == 6 || sides == 8 || sides == 10 || sides == 12 || sides == 20) {
				sb.append("[image \"path=/dice/small/d"+sides+".png\"]");
				appendRollValue(sb, "", result, "");
			} else {
				appendRollValue(sb, "<", result, ">");
			}
			sb.append("]");
		}
		
		if(results.size() > 1) sb.append(")");
		return new Result(value, sb.toString(), hadCriticalFailure, hadCriticalSuccess);
	}

	private void appendRollValue(StringBuilder sb, String prefix, DiceResult result, String postfix) {
		sb.append("[group ");
		sb.append(prefix);
		boolean changedStyle = false;
		if(result.criticalFailure) {
			sb.append("[style \"color=#FF0000\"]");
			changedStyle = true;
		} else if(result.criticalSuccess) {
			sb.append("[style \"color=#008800\"]");
			changedStyle = true;
		}
		sb.append(result.value);
		if(changedStyle) {
			sb.append("[style \"color=#000000\"]");
		}
		sb.append(postfix);
		sb.append("]");
	}
	
	private final class DiceResult implements Comparable<DiceResult> {
		private final int value;
		private final boolean criticalFailure;
		private final boolean criticalSuccess;
		private final boolean counted;
		
		private boolean dropped = false;
		private boolean kept = true;
		
		public DiceResult(int value, boolean counted) {
			this.value = value;
			this.criticalFailure = criticalFailureCondition.matches(value);
			this.criticalSuccess = criticalSuccessCondition.matches(value);
			this.counted = counted;
		}
		
		public boolean shouldBeIncluded() {
			return counted && !dropped && kept;
		}

		@Override
		public int compareTo(DiceResult other) {
			return value - other.value;
		}
	}
	
	public static enum Action {
		NOTHING,
		DROP,
		KEEP;
	}
}

File: core/common/scripting/expression/dice.js
import { Expression } from '../expression.js';
import { Result } from '../result.js';
import { Condition } from './condition.js';
import { ZERO, ONE, Expr } from './expr.js';
import { DiceRoller } from '../../util/dice-roller.js';

class DiceResult {
    sides;
    value;
    criticalFailure;
    criticalSuccess;
    counted;

    dropped = false;
    kept = true;

    constructor(sides, value, criticalFailure, criticalSuccess, counted) {
        this.sides = sides;
        this.value = value;
        this.criticalFailure = criticalFailure;
        this.criticalSuccess = criticalSuccess;
        this.counted = counted;
    }

    shouldBeIncluded() {
        return this.counted && !this.dropped && this.kept;
    }
}

export class Dice extends Expression {
    count;
    sides;

    criticalFailureCondition;
    criticalSuccessCondition;

    rerollCondition;
    maxRerollCount = 100;

    explodeDice = false;
    maxExplodeDice = 100;

    lowAction = Dice.Action.NOTHING;
    lowCount = ZERO;
    highAction = Dice.Action.NOTHING;
    highCount = ZERO;

    label = '';

    constructor(count, sides) {
        super();

        this.count = count;
        this.sides = sides;
    }

    setCriticalFailureCondition(criticalFailureCondition) {
        this.criticalFailureCondition = criticalFailureCondition;
    }

    setCriticalSuccessCondition(criticalSuccessCondition) {
        this.criticalSuccessCondition = criticalSuccessCondition;
    }

    setRerollCondition(rerollCondition) {
        this.rerollCondition = rerollCondition;
    }

    setMaxRerollCount(maxRerollCount) {
        this.maxRerollCount = maxRerollCount;
    }

    setExplodeDice(explodeDice) {
        this.explodeDice = explodeDice;
    }

    setMaxExplodeCount(maxExplodeCount) {
        this.maxExplodeCount = maxExplodeCount;
    }

    setLowAction(lowAction) {
        this.lowAction = lowAction;
    }

    setLowCount(lowCount) {
        this.lowCount = lowCount;
    }

    setHighAction(highAction) {
        this.highAction = highAction;
    }

    setHighCount(highCount) {
        this.highCount = highCount;
    }

    setLabel(label) {
        this.label = label;
    }

    eval(context) {
        // eval basic values 
		//---------------------------------------------------
        // (count + sides)
        const countValue = Math.trunc(this.count.eval(context).getValue());
        const sidesValue = Math.trunc(this.sides.eval(context).getValue());
        if(countValue == 0) return ZERO;
        if(countValue <= 0) throw new Error(`Cannot throw ${countValue} dice`);
        if(sidesValue <= 0) throw new Error(`Cannot throw ${sidesValue} sided dice`);

        // (high/low counts)
        const lowCountValue = Math.trunc(this.lowCount.eval(context).getValue());
        const highCountValue = Math.trunc(this.highCount.eval(context).getValue());

        // prepare default conditions if none are set
        if(!this.criticalFailureCondition) this.criticalFailureCondition = new Condition(Condition.Type.EQUAL, ONE);
        if(!this.criticalSuccessCondition) this.criticalSuccessCondition = new Condition(Condition.Type.EQUAL, new Expr(c => new Result(sidesValue, String(sidesValue))));

        // eval condition values
        this.criticalFailureCondition.eval(context);
        this.criticalSuccessCondition.eval(context);
        if(this.rerollCondition) this.rerollCondition.eval(context);

		// calculate all required dice
        //---------------------------------------------------
        var results = [];
        var realCount = 0;
        for(var i=0; i<countValue; i++) {
            var roll;
            var explodeCount = 0;
            do {
                // calculate roll
                roll = DiceRoller.roll(sidesValue);

                realCount++;
                if(realCount > 1000) throw new Error(`Number of rolls exceeded 1000`);

                // handle re-rolling
                var rerollCount = 0;
                while(this.rerollCondition && this.rerollCondition.matches(roll) && rerollCount < this.maxRerollCount) {
                    results.push(new DiceResult(sidesValue, roll, this.criticalFailureCondition.matches(roll), this.criticalSuccessCondition.matches(roll), false));
                    roll = DiceRoller.roll(sidesValue);
                    rerollCount++;

                    realCount++;
                    if(realCount > 1000) throw new Error(`Number of rolls exceeded 1000`);
                }

                results.push(new DiceResult(sidesValue, roll, this.criticalFailureCondition.matches(roll), this.criticalSuccessCondition.matches(roll), true));
			// handle exploding dice
            } while(this.explodeDice && roll == sidesValue && explodeCount++ < this.maxExplodeCount);
        }
        
		// apply drop/keep modifiers
        //---------------------------------------------------
        var sortedResults = Array.from(results);
        sortedResults.sort((a, b) => a.value - b.value);
        var currentLowCount = 0;
		for(var i=0; i<sortedResults.length; i++) {
            const result = sortedResults[i];
            if(result.counted) {
                if(this.lowAction == Dice.Action.DROP && currentLowCount < lowCountValue) {
                    result.dropped = true;
                } else if(this.lowAction == Dice.Action.KEEP && currentLowCount >= lowCountValue) {
                    result.kept = false;
                }
                currentLowCount++;
            }
        }
        var currentHighCount = 0;
		for(var i=sortedResults.length-1; i>=0; i--) {
            const result = sortedResults[i];
            if(result.counted) {
                if(this.highAction == Dice.Action.DROP && currentHighCount < highCountValue) {
                    result.dropped = true;
                } else if(this.highAction == Dice.Action.KEEP && currentHighCount >= highCountValue) {
                    result.kept = false;
                }
            }
            currentHighCount++;
        }

		// calculate result
        //---------------------------------------------------
        var value = 0;
        var string = '';
        var diceRolls = [];

        if(results.length > 1) string = string + '(';
        for(var i=0; i<results.length; i++) {
            const result = results[i];

            // apply value
            if(result.shouldBeIncluded()) {
                value += result.value;
            }

            // track rolls //TODO: convert this to an actual class - maybe just use DiceResult?
            diceRolls.push({
                sides: result.sides,
                result: result.value,
                label: this.label,
                counted: result.shouldBeIncluded(),
                cf: result.criticalFailure,
                cs: result.criticalSuccess
            });

            // build string (with colored misses/crits)
            if(i > 0) string = string + '+';
            if(sidesValue == 4 || sidesValue == 6 || sidesValue == 8 || sidesValue == 10 || sidesValue == 12 || sidesValue == 20) {
                string = string + this.getRollValue('', result, '', `chat-dice-bg chat-dice-${sidesValue}`);
            } else {
                string = string + this.getRollValue('&lt;', result, '&gt;', '');
            }
        }
        if(results.length > 1) string = string + ')';

        return new Result(value, string, diceRolls);
    }

    getRollValue(prefix, result, postfix, cl) {
        // build class name(s)
        if(result.criticalFailure) {
            cl = (cl === '' ? 'chat-dice-fail' : cl + ' chat-dice-fail')
        } else if(result.criticalSuccess) {
            cl = (cl === '' ? 'chat-dice-crit' : cl + ' chat-dice-crit')
        }

        // build string
        var string = prefix;
        string = string + `<span class="${cl}">`;
        string = string + result.value;
        string = string + '</span>';
        string = string + postfix;

        return string;
    }

    static Action = {
        NOTHING: 'NOTHING',
        DROP: 'DROP',
        KEEP: 'KEEP'
    }
}
